### PR TITLE
Add workflow for checking sidecar digests

### PR DESCRIPTION
**Pull request by mkuznyetsov**
Signed-off-by: Mykhailo Kuznietsov <mkuznets@redhat.com>

### What does this PR do?
Add regularly scheduled workflow for checking sidecar image digests in base_images file. If new digest is detected, a pull request with it's update will be generated

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: .github/workflows/sidecar-check.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+#
+# SPDX-License-Identifier: EPL-2.0
+#
+
+name: Sidecar Digest validation
+
+on:
+  workflow_dispatch:
+  schedule:
+  - cron: "0 0 * * *"
```
Comment (nickboldt) [inline .github/workflows/sidecar-check.yml:15]: is every day at midnight a good time for this? Just wondering what other processes might be running in parallel ... prolly ok, depending on which TZ this is.

File: .github/workflows/sidecar-check.yml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        name: Login to docker.io
+        uses: docker/login-action@v1
+        with:
+          registry: docker.io
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_PASSWORD }}
+      - name: Install skopeo
+        run: |
+          . /etc/os-release
+          echo "deb https://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/xUbuntu_${VERSION_ID}/ /" | sudo tee /etc/apt/sources.list.d/devel:kubic:libcontainers:stable.list
```
Comment (nickboldt) [inline .github/workflows/sidecar-check.yml:50]: wow, we're using opensuse + ubuntu here? shame we can't use a single OS like UBI8 :)

**Review by nickboldt — APPROVED** [review]
LGTM, let's merge it and see what happens.

File: .github/workflows/sidecar-check.yml (hunk 2 above)
Comment (mkuznyetsov) [reply .github/workflows/sidecar-check.yml:50]: well looks like there are no runners with ubuntu 20.10 yet, that have skopeo repos configured in them
so I guess it's either this, or running skopeo from some docker container (or using curl instead?)

File: arbitrary-users-patch/check_sidecar_image_digests.sh [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    set -e
+}
+
+BUILT_IMAGES=""
+while read -r line; do
+  dev_container_name=$(echo "$line" | tr -s ' ' | cut -f 1 -d ' ')
+  base_image_name=$(echo "$line" | tr -s ' ' | cut -f 2 -d ' ')
+  base_image_digest=$(echo "$line" | tr -s ' ' | cut -f 3 -d ' ' )
+  echo "Checking ${NAME_FORMAT}/${dev_container_name} based on $base_image_name ..."
+  latest_digest="$(skopeo inspect --tls-verify=false docker://"${base_image_name}" 2>/dev/null | jq -r '.Digest')"
```
Comment (benoitf) [inline arbitrary-users-patch/check_sidecar_image_digests.sh:48]: I think we can just use curl there for digest no ? (no need to have skopeo)

File: .github/workflows/sidecar-check.yml (hunk 2 above)
Comment (nickboldt) [reply .github/workflows/sidecar-check.yml:50]: https://github.com/redhat-developer/codeready-workspaces/blob/crw-2-rhel-8/product/util.groovy#L353-L361 is another way to get skopeo from the official RH released container registry.redhat.io/rhel8/skopeo

File: arbitrary-users-patch/check_sidecar_image_digests.sh (hunk 3 above)
Comment (nickboldt) [reply arbitrary-users-patch/check_sidecar_image_digests.sh:48]: HOWEVER... this is way more complicated as it requires doing some fancy footwork w/ docker authentication to be able to fetch manifests/digests. If the world was all quay, it'd be simple.

```
  # for quay, no need to authenticate
  AUTH=""
  # for docker, must authenticate with Bearer token
  if [[ ${base_image_name} == *"docker.io"* ]]; then
    REPOSITORY=${base_image_name#*/}; REPOSITORY=${REPOSITORY%:*} # get antora/antora
    TOKEN=$(curl -s "https://auth.docker.io/token?service=registry.docker.io&scope=repository:$REPOSITORY:pull" | jq -r .token)
    AUTH='-D - -H "Authorization: Bearer '$TOKEN'"'
    manifest_URL=${base_image_name/:/\/manifests\/}; manifest_URL=https://index.${manifest_URL/\//\/v2\/}
  else
    manifest_URL=${base_image_name/:/\/manifests\/}; manifest_URL=https://${manifest_URL/\//\/v2\/}
  fi
  manifest="$(curl -s ${AUTH} -H 'Accept: application/vnd.docker.distribution.manifest.v2+json' ${manifest_URL})"

  if [[ ${base_image_name} == "docker.io"* ]]; then
    latest_digest="sha256:$(echo $manifest | jq '.Docker-Content-Digest' -r)"
  else 
    latest_digest="sha256:$(echo $manifest | sha256sum | cut -f 1 -d ' ')"
  fi
  echo "latest digest ---> ${latest_digest}"
```

^^ that sort of works, but also doesn't :(

**Review by nickboldt — APPROVED** [review]

File: .github/workflows/sidecar-check.yml [hunk 4]
```diff
@@ -0,0 +1,59 @@
+#
+# Copyright (c) 2020-2021 Red Hat, Inc.
+# This program and the accompanying materials are made
+# available under the terms of the Eclipse Public License 2.0
+# which is available at https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+#
+
```
Comment (nickboldt) [inline .github/workflows/sidecar-check.yml:9]: please add a description/purpose for this script so future maintainers will know why we're doing this

**Review by nickboldt — CHANGES_REQUESTED** [review]
please add a description/purpose for this script so future maintainers will know why we're doing this (around line 9)

File: arbitrary-users-patch/check_sidecar_image_digests.sh (hunk 3 above)
Comment (mkuznyetsov) [reply arbitrary-users-patch/check_sidecar_image_digests.sh:48]: yeah, I thought in this case it is probably simpler to defer complex logic to skopeo

**Comment by nickboldt** [comment]
Follow up work in https://github.com/eclipse/che/issues/19695

If you add a comment to the top of the file explaining the purpose here, then we should be good to merge this as is.

@benoitf @ericwill do you agree?

**Comment by ericwill** [comment]
> Follow up work in [eclipse/che#19695](https://github.com/eclipse/che/issues/19695)
> 
> If you add a comment to the top of the file explaining the purpose here, then we should be good to merge this as is.
> 
> @benoitf @ericwill do you agree?

Agreed

**Review by nickboldt — APPROVED** [review]

**Pull request merged**